base_layer/common_types/src/tari_address/dual_address.rs

```diff
@@ -0,0 +1,418 @@
+// Copyright 2020. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+use tari_common::configuration::Network;
+use tari_crypto::tari_utilities::ByteArray;
+use tari_utilities::hex::{from_hex, Hex};
+
+use crate::{
+    dammsum::{compute_checksum, validate_checksum},
+    emoji::{EMOJI, REVERSE_EMOJI},
+    tari_address::{TariAddressError, TariAddressFeatures, INTERNAL_DUAL_SIZE},
+    types::PublicKey,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
+pub struct DualAddress {
+    network: Network,
+    features: TariAddressFeatures,
+    public_view_key: PublicKey,
+    public_spend_key: PublicKey,
+}
+
+impl DualAddress {
+    /// Creates a new Tari Address from the provided public keys, network and features
+    pub fn new(
+        view_key: PublicKey,
+        spend_key: PublicKey,
+        network: Network,
+        features: TariAddressFeatures,
+    ) -> DualAddress {
+        Self {
+            network,
+            features,
+            public_view_key: view_key,
+            public_spend_key: spend_key,
+        }
+    }
```

Comment:
Dual address with interactive only features should not be allowed.
```suggestion
    pub fn new(
        view_key: PublicKey,
        spend_key: PublicKey,
        network: Network,
        features: TariAddressFeatures,
    ) -> Result<DualAddress, TariAddressError> {
        if features == TariAddressFeatures::create_interactive_only() {
            return Err(TariAddressError::InvalidFeatures);
        }
        Ok(Self {
            network,
            features,
            public_view_key: view_key,
            public_spend_key: spend_key,
        })
    }
```

Reply:
Nope can be valid. Why do you want this blocked?

Reply:
Dual without one-sided is pointless, it will only add comms overhead.

Reply:
What comms overhead?

And what about the use case where you have a wallet, only want 1-sided transaction but you want to give out your view key to a thrid party?